Main.hs

```diff
@@ -2,32 +2,142 @@ module Main where
 
 import Control.Monad (unless)
 import Text.Printf (printf)
+import Data.Either
 
-data Expr
+data Expr =
+  Num Double
+  | Sqrt Expr
+  | Add Expr Expr
+  | Sub Expr Expr
+  | Mul Expr Expr
+  | Div Expr Expr
+  | Pow Expr Expr
 
 instance Show Expr where
-  show = undefined
+  show (Num a) = show a
+  show (Sqrt a) = show "sqrt(" ++ show a ++ ")" 
+  show (Add x y) = "(" ++ show x ++ " + " ++ show y ++ ")"
+  show (Sub x y) = "(" ++ show x ++ " - " ++ show y ++ ")"
+  show (Mul x y) = "(" ++ show x ++ " * " ++ show y ++ ")"
+  show (Div x y) = "(" ++ show x ++ " / " ++ show y ++ ")"
+  show (Pow x y) = "(" ++ show x ++ " ^ " ++ show y ++ ")"
 
 instance Eq Expr where
-  (==) = undefined
+  (==) (Num a) (Num b) = a == b
+  (==) (Sqrt a) (Sqrt b) = a == b
+  (==) (Add a b) (Add c d) = (a == c) && (b == d)
+  (==) (Sub a b) (Sub c d) = (a == c) && (b == d)
+  (==) (Mul a b) (Mul c d) = (a == c) && (b == d)
+  (==) (Div a b) (Div c d) = (a == c) && (b == d)
+  (==) (Pow a b) (Pow c d) = (a == c) && (b == d)
+  (==) _ _ = False
 
-data Error
+
+data Error = DivZero Expr | NegRoot Expr
 
 instance Show Error where
-  show = undefined
+  show (DivZero expr) = "division by zero is forbidden " ++ show expr
+  show (NegRoot expr) = "sqrt of a negative number is forbidden " ++ show expr
 
 instance Eq Error where
-  (==) = undefined
+  (==) x y = show x == show y
+
+evenRoots = [1/x | x<-[2,4..31]]  -- we can take more here but I think 1e-9 is pretty small enough
+
+performOperation :: (Double -> Double -> Double) -> Expr -> Expr -> Either Error Double
+performOperation func a b
+  | isLeft x = x
+  | isLeft y = y
+  | otherwise = Right (func (fromRight 1 x) (fromRight 1 y))
+  where 
+    x = eval a
+    y = eval b
+
 
 eval :: Expr -> Either Error Double
-eval = undefined
+eval (Num a) = Right a
+eval (Sqrt a) = case eval a of
+  (Left x) -> Left x
+  (Right x) -> if x < 0 then Left (NegRoot (Sqrt a)) else Right (sqrt x) 
+eval (Add a b) = performOperation (+) a b
+eval (Sub a b) = performOperation (-) a b
+eval (Mul a b) = performOperation (*) a b
+eval (Div a b) = 
+  if fromRight 1 (eval b) == 0 
+    then Left (DivZero (Div a b)) 
+  else 
+    performOperation (/) a b  
+
+-- I needed to handle myself negative roots because for negative bases since it uses logarithms for the calculations gives nan
+eval (Pow a b) 
+  | isLeft x = x
+  | isLeft y = y  
+  | fromRight 0 x < 0 && fromRight 0 y < 1 =
+    if fromRight 0 y `elem` evenRoots
+      then Left (NegRoot (Pow a b)) 
+    else   --make it positive first then take the power and then multiply it again
+      Right (negate (abs (fromRight 0 x) ** fromRight 0 y))
+  | otherwise = performOperation (**) a b
+  where 
+    x = eval a
+    y = eval b
+
+
+{-  Why this doesn't work?, somehow for negative numbers and odd powers it gives NaN. It would be nice to have an 
```

Comment:
The power function for floating point numbers is finicky. For some combination of numbers, it's just not defined. And I believe it's not about even roots: 

```
ghci> (-13) ** (1/9)
Na
```